File: datasets/Part 2 - Regression/random_forest_regression.py
# -*- coding: utf-8 -*-
"""
Created on Sun May  8 23:19:27 2022

@author: mudar
"""

# Regresion con Random Forest

# Como importar las librerias
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Importar el dataset
dataset = pd.read_csv('data/Position_Salaries.csv')
X = dataset.iloc[:, 1:2].values # para que no sea un vector y sea una matriz ponemos 1:2
y = dataset.iloc[:, 2].values


# Dividir el dataset en conjunto de entrenamiento y conjunto de testing
# 80% conjunto entrenamiento y 20% conjunto de testing
"""
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)
"""


# Escalado de variables
# Para que no haya tanta diferencia en los rangos de
# valores en los datos de las distintas columnas
"""
from sklearn.preprocessing import StandardScaler
sc_X = StandardScaler()
X_train = sc_X.fit_transform(X_train)
X_test = sc_X.transform(X_test)
"""


# Ajustar el Random Forest con el dataset
from sklearn.ensemble import RandomForestRegressor
regression = RandomForestRegressor(n_estimators = 300, random_state = 0)
regression.fit(X, y)


# Prediccion de nuestros modelos con Random Forest
y_pred = regression.predict([[6.5]])


# Visualizacion de los resultados del Random Forest
X_grid = np.arange(min(X), max(X), 0.01)
X_grid = X_grid.reshape(len(X_grid), 1)
plt.scatter(X, y, color = 'red')
plt.plot(X_grid, regression.predict(X_grid), color = 'blue')
plt.title('Modelo de Regresion con Random Forest')
plt.xlabel('Posicion del empleado')
plt.ylabel('Sueldo (en $)')
plt.show()

File: datasets/Part 2 - Regression/random_forest_regression.R

# Random Forest Regression

setwd("~/repos/machinelearning-az/datasets/Part 2 - Regression")

# Importar el dataset
dataset = read.csv('data/Position_Salaries.csv')
dataset = dataset[, 2:3]

# En este caso no se va a dividir en conjunto de entrenamiento y test
# debido a la escasa cantidad de datos

# Dividir los datos en conjunto de entrenamiento y conjunto de test
# library(caTools) # Instalar con install.packages('caTools')
# set.seed(123) # Para que salgan los mismos resultados que en el curso
# split = sample.split(dataset$Purchased, SplitRatio = 0.8)
# training_set = subset(dataset, split == TRUE)
# testing_set = subset(dataset, split == FALSE)


# Escalado de valores
#training_set[, 2:3] = scale(training_set[, 2:3])
#testing_set[, 2:3] = scale(testing_set[, 2:3])


# Ajustar Modelo de Random Forest con el Conjunto de Datos
library(randomForest) # Instalar con install.packages('randomForest')
set.seed(1234)
# x = dataset[1] para que lo cuente como data frame (debe ser asi con esta funcion)
regression = randomForest(x = dataset[1],
                          y = dataset$Salary,
                          ntree = 500) 

# Prediccion de nuevos resultados con Regresion
y_pred = predict(regression, newdata = data.frame(Level = 6.5))

# Visualizacion del Modelo con Random Forest
library(ggplot2) # Instalar con install.packages('ggplot2')
x_grid = seq(min(dataset$Level), max(dataset$Level), 0.01)
ggplot() +
  geom_point(aes(x = dataset$Level, y = dataset$Salary),
             color = 'red') +
  geom_line(aes(x = x_grid, y = predict(regression,
                                        newdata = data.frame(Level = x_grid))),
            color = 'blue') +
  ggtitle('Prediccion (Random Forest)') +
  xlab('Nivel del empleado') +
  ylab('Sueldo (en $)')
